[AAE-22377] Implemented max items limit per request

RCorsaletti opened the pull request:
STILL WP need parameters for max and enabled/disabled

RCorsaletti committed:
[AAE-22377] Implemented max items limit per request

RCorsaletti committed:
[AAE-22377] Fixed tests and Sonar compliance

RCorsaletti committed:
[AAE-22377] Fixed tests

graric commented on activiti-cloud-service-common/activiti-cloud-services-dbp-rest/src/main/java/org/activiti/cloud/alfresco/argument/resolver/AlfrescoPageArgumentMethodResolver.java:
```diff
@@ -29,6 +30,9 @@ public class AlfrescoPageArgumentMethodResolver implements PageableArgumentResol
     private final AlfrescoPageParameterParser pageParameterParser;
     private final PageableHandlerMethodArgumentResolver pageableHandlerMethodArgumentResolver;
 
+    private static final int MAX_ITEMS_LIMIT = 1000; // add parameter
+    private static final boolean MAX_ITEMS_LIMIT_ENABLED = true; // add parameter
```
Are these comments needed?

erdemedeiros replied:
I think both values should use system properties, so that they can be overridden if needed. I'd also use `false` as default value for `MAX_ITEMS_LIMIT_ENABLED`.

graric commented on activiti-cloud-service-common/activiti-cloud-services-dbp-rest/src/main/java/org/activiti/cloud/alfresco/argument/resolver/AlfrescoPageArgumentMethodResolver.java:
```diff
@@ -58,6 +62,15 @@ public Pageable resolveArgument(
+            !alfrescoQueryParameters.getMaxItemsParameter().isSet() ||
+            alfrescoQueryParameters.getMaxItemsParameter().getValue() > MAX_ITEMS_LIMIT;
+
+        if (isMaxItemsExceedingLimit && MAX_ITEMS_LIMIT_ENABLED) {
```
What if we move these lines into a method and just call it directly in the if? 
That way we can make it more readable IMO

erdemedeiros replied:
+1

graric commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/rest/TaskEntityControllerIT.java:
```diff
@@ -114,7 +114,7 @@ void setUp() {
 
     @Test
-    public void findAllShouldReturnAllResultsUsingAlfrescoMetadataWhenMediaTypeIsApplicationJson() throws Exception {
+    void findAllShouldReturnAllResultsUsingAlfrescoMetadataWhenMediaTypeIsApplicationJson() throws Exception {
```
Can you please follow the test naming convention? 🙏 

https://hyland.atlassian.net/wiki/spaces/HXP/pages/1308014107/Studio+-+Coding+Style+Guidelines#Naming-and-other-test-conventions

graric replied:
Also, the public access modifier is no longer needed in JUnit 5 tests, which is great news!

erdemedeiros commented on activiti-cloud-service-common/activiti-cloud-services-dbp-rest/src/main/java/org/activiti/cloud/alfresco/argument/resolver/AlfrescoPageArgumentMethodResolver.java:
```diff
@@ -58,6 +62,15 @@ public Pageable resolveArgument(
+            alfrescoQueryParameters.getMaxItemsParameter().getValue() > MAX_ITEMS_LIMIT;
+
+        if (isMaxItemsExceedingLimit && MAX_ITEMS_LIMIT_ENABLED) {
+            Pageable limitedPageable = PageRequest.of(0, MAX_ITEMS_LIMIT);
+            return new AlfrescoPageRequest(MAX_ITEMS_LIMIT + 1L, MAX_ITEMS_LIMIT, limitedPageable);
```
IMO, silently changing the values provided via query param sounds confusing to me and I think people will report bugs for this. Maybe it's better to raise an exception in this case.

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/rest/TaskEntityControllerIT.java:
```diff
@@ -250,4 +250,60 @@ public void should_returnTaskPermissions_when_invokeGetTaskById() throws Excepti
+            .willReturn(new PageImpl<>(Collections.singletonList(buildDefaultTask()), pageRequest, 1000));
+
+        //when
+        MvcResult result = mockMvc
+            .perform(get("/v1/tasks?skipCount=2001&maxItems=2000").accept(MediaType.APPLICATION_JSON))
+            //then
+            .andExpect(status().isOk())
+            .andReturn();
+
+        assertThatJson(result.getResponse().getContentAsString())
+            .node("list.pagination.skipCount")
+            .isEqualTo(1001)
+            .node("list.pagination.maxItems")
+            .isEqualTo(1000)
+            .node("list.pagination.count")
+            .isEqualTo(1)
+            .node("list.pagination.hasMoreItems")
+            .isEqualTo(false)
+            .node("list.pagination.totalItems")
+            .isEqualTo(1002);
+    }
```
In any case, we should never touch to `skipCount` that gives how many elements should be left behind before getting the first element to be returned (this is the equivalent of start index). In this specific case, the caller is asking for up to `2000` elements, skipping the first 2001 elements from the results, but what he's getting in return is up `1000` (what could still be fine), skipping only the first 1001 elements from the results: he's getting 1000 elements that he's not expecting.

erdemedeiros reviewed (changes requested):
Hello, I think this change is going to be confusing for end users. Let's discuss about it.

RCorsaletti committed:
[AAE-22377] Fixed tests and skip count logic

RCorsaletti committed:
Merge branch 'develop' into feature/AAE-22377-resolver-for-pagination-limit

alfresco-build committed:
chore(deps): upgrade Activiti/Activiti to version 8.5.0-alpha.4 (#1442)

from: https://github.com/Activiti/Activiti

Co-authored-by: alfresco-build <>

tom-dal committed:
AAE-22075 Fix "detaching an uninitialized collection" logs and improve task search query performance (#1441)

* AAE-22075 Wip initial commit

* WIP

* AAE-22075 Wip fixed failing test

* AAE-22075 reproduce error. ("spring.jpa.properties.hibernate.enable_lazy_load_no_trans=true", if set to false, exception is raised)

* AAE-22075 reproduce exact situation of issue

* AAE-22075 tentative with DSL filters

* Revert "AAE-22075 tentative with DSL filters"

This reverts commit 6a148ac4187d6741f50bab7dee0582655beb3fb8.

* AAE-22075 test with entity graph

* AAE-18428 Fix Swagger API suggest to use the HAL-JSON format (#1306)

* AAE-21097 Implement permission-based authorization (#1427)

* AAE-21097 add permission getter on Jwt adapter, implement conversion to granted authority and include permissions in auth flow

* AAE-21097 add AuthorizationConfigurerIT, fix omitted methods logic

* AAE-21097 fix tests

* AAE-21097 add license header

* AAE-21097 remove unused import

* AAE-21097 CustomAuthorizationManager implemented with different authority types for role and permission

* AAE-21097 move CustomAuthorizationManager

* AAE-21097 license header

* AAE-22123 Fix NPE when getting permissions (#1431)

* AAE-22123 added key check in KeycloakResourceJwtAdapter

* AAE-22123 change test

* chore(deps): upgrade Activiti/Activiti to version 8.5.0-alpha.2 (#1432)

from: https://github.com/Activiti/Activiti

Co-authored-by: alfresco-build <>

* build(deps): bump github/codeql-action from 3.23.1 to 3.25.3 (#1429)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.23.1 to 3.25.3.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/0b21cf2492b6b02c465a3e5d7c473717ad7721ba...d39d31e687223d841ef683f52467bd88e9b21c14)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* AAE-22123 Fix omitted methods (#1433)

* AAE-22123 fix omitted methods management

* AAE-22123 fix omitted methods management and tests

* AAE-22075 perform single queries on join table after select on task table

* Revert "AAE-22075 perform single queries on join table after select on task table"

This reverts commit 3e3576395b42ba2eb61b370864797512b8e4bce6.

* AAE-22075 use jpaquery and fluentyquery

* AAE-22075 reduced number of queries, applied sorting parameters

* AAE-22075 fix sorting

* AAE-22075 fix sorting

* AAE-22075 fix

* AAE-22075 remove unused imports

* AAE-22075 remove sql logs from test

* AAE-22075 fix import

* AAE-22075 added tests

* AAE-22075 refactoring

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Akhilesh Pamidimarthi <akhilesh.pamidimarthi@hyland.com>
Co-authored-by: jesty <davidecerbo@gmail.com>
Co-authored-by: RahulKumar1606 <126051097+RahulKumar1606@users.noreply.github.com>
Co-authored-by: Alfresco Build <8039454+alfresco-build@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mavotto committed:
AAE-22648 Upgrade spring-boot version to fix vulnerability (#1444)

AAE-22648 upgrade spring boot version to fix vulnerability

alfresco-build committed:
Update to version 8.6.0-SNAPSHOT [skip ci]

RCorsaletti committed:
[AAE-22377] Exception if user exceed pagination limit

RCorsaletti committed:
[AAE-22377] Exception if user exceed pagination limit
